ui-manchette-with-spacetimechart/src/hooks/useManchetteWithSpaceTimeChart.ts

```diff
@@ -0,0 +1,231 @@
+    scales: [],
+  });
+
+  const { xZoom, yZoom, xOffset, yOffset, panY, panning, scrollPosition, isProportional } = state;
+
+  const paths = usePaths(projectPathTrainResult, selectedProjection);
+
+  const checkOverflow = useCallback((isOverflowFromCallback: boolean) => {
+    setState((prev) => ({ ...prev, panY: isOverflowFromCallback }));
+  }, []);
+  useIsOverflow(manchette, checkOverflow);
+
+  // Memoize timeWindow to avoid recalculation on each render
+  const timeWindow = useMemo(
+    () => computeTimeWindow(projectPathTrainResult),
+    [projectPathTrainResult]
+  );
+
+  // Memoize computedOperationalPoints to avoid recalculations unless the dependencies change
+  const operationalPointsToDisplay = useMemo(
+    () => calcOperationalPointsToDisplay(operationalPoints, { height, isProportional, yZoom }),
+    [operationalPoints, height, isProportional, yZoom]
+  );
+  const operationalPointsWithHeight = useMemo(
+    () =>
+      calcOperationalPointsHeight(operationalPointsToDisplay, { height, isProportional, yZoom }),
+    [operationalPointsToDisplay, height, yZoom, isProportional]
+  );
+
+  const operationalPointsWithPosition = useMemo(
+    () => getOperationalPointsWithPosition(operationalPointsToDisplay),
+    [operationalPointsToDisplay]
+  );
+
+  const zoomYIn = useCallback(() => {
+    if (yZoom < MAX_ZOOM_Y) {
+      setState((prev) => ({ ...prev, yZoom: yZoom + ZOOM_Y_DELTA }));
+    }
+  }, [yZoom]);
+
+  const zoomYOut = useCallback(() => {
+    if (yZoom > MIN_ZOOM_Y) {
+      setState((prev) => ({ ...prev, yZoom: yZoom - ZOOM_Y_DELTA }));
+    }
+  }, [yZoom]);
+
+  const handleScroll = useCallback(() => {
+    if (!isShiftPressed && manchette.current) {
+      const { scrollTop } = manchette.current;
+      if (scrollTop || scrollTop === 0) {
+        setState((prev) => ({ ...prev, scrollPosition: scrollTop, yOffset: scrollTop }));
+      }
+    }
+  }, [isShiftPressed, manchette]);
+
+  const handleKeyDown = useCallback((event: KeyboardEvent) => {
+    if (event.key === 'Shift') {
+      setIsShiftPressed(true);
+    }
+  }, []);
+
+  const handleKeyUp = useCallback((event: KeyboardEvent) => {
+    if (event.key === 'Shift') {
+      setIsShiftPressed(false);
+    }
+  }, []);
+
+  useEffect(() => {
+    window.addEventListener('keydown', handleKeyDown);
+    window.addEventListener('keyup', handleKeyUp);
+
+    return () => {
+      window.removeEventListener('keydown', handleKeyDown);
+      window.removeEventListener('keyup', handleKeyUp);
+    };
+  }, [handleKeyDown, handleKeyUp]);
+
+  const toggleMode = useCallback(() => {
+    setState((prev) => ({ ...prev, isProportional: !prev.isProportional }));
+  }, []);
+
+  const computedScales = useMemo(
+    () => getScales(operationalPointsWithPosition, { height, isProportional, yZoom }),
+    [operationalPointsWithPosition, height, isProportional, yZoom]
+  );
+
+  // Memoize the returned object to avoid unnecessary recalculations
+  return useMemo(
```

Comment:
My concern remains here: Since it's all memoized together, a component depending on `manchetteProps` (typically, the manchette) will be rerendered when `spaceTimeChartProps` is updated, and vice versa.

The two big branches of what you return from that function (`manchetteProps` and `spaceTimeChartProps`) should be memoized separately.

Something like:

```typescript
const manchetteProps = useMemo(/* ... */);
const spaceTimeChartProps = useMemo(/* ... */);

return {
  manchetteProps,
  spaceTimeChartProps,
  /* ... */
};
```

Reply:
done